<a href="https://colab.research.google.com/github/JitendraAlim/Data-Science-Toolkit/blob/master/Python/Pandas/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas

![Python & Pandas](https://geo-python.github.io/2017/_images/pandas_logo.png "Python & Pandas")

By - Jitendra Alim

***
# Importing Required Libraries

In [102]:
# Pandas for Data Management
import pandas as pd

# Pandas Profiling for HTML Summary
import pandas_profiling as pr

# Numpy for Calculations
import numpy as np

# Datetime for Time and Date Operations
import datetime as dt

# Matplot and Seaborn for Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Scipy for Scientific Operations
import scipy as sp

***
# Today

In [103]:
print("Last Updated On: ", dt.datetime.now())

Last Updated On:  2020-09-25 14:15:05.234076


***
# Pandas Version

In [104]:
pd.__version__

'1.0.5'

***
# Version of Libraries in Python

In [105]:
# pd.show_versions()

***
# Setting the Display Options

In [106]:
# For displaying many rows as you want
pd.options.display.max_rows = 100

# For displaying many columns as you want
pd.options.display.max_columns = 100

# For displaying numbers with comma seperator
pd.options.display.float_format = '{0:,.2f}'.format

# For expanding the frame for output
pd.options.display.expand_frame_repr = True

# For aligning the column headers to left
pd.options.display.colheader_justify = 'left'

# To get exact memor usage of dataframes in info()
# pd.options.display.memory_usage = 'deep'

One can use any of the required display options.

***
# Resetting the Display Options

In [107]:
# pd.reset_option('display')

This above command will reset all the applied display options to default.

***
# Reading online data into Python

In [108]:
titanic = pd.read_csv('http://bit.ly/titanic_passengers')
print('Dimension of Titanic :', titanic.shape)

Dimension of Titanic : (891, 12)


In [109]:
alcohol = pd.read_csv('http://bit.ly/alcohol_consumed')
print('Dimension of Alcohol :', alcohol.shape)

Dimension of Alcohol : (193, 6)


In [110]:
ufo = pd.read_csv('http://bit.ly/ufo_sights')
print('Dimension of UFO :', ufo.shape)

Dimension of UFO : (5864, 5)


In [111]:
iris = pd.read_csv('http://bit.ly/iris_samples')
print('Dimension of Iris :', iris.shape)

Dimension of Iris : (150, 6)


In [112]:
bigmart = pd.read_csv('http://bit.ly/big_mart')
print('Dimension of Big Mart :', bigmart.shape)

Dimension of Big Mart : (2444, 12)


<p> These were some of the datasets that i've stored on an online repository </p>

***
# Stored Variables

In [113]:
%whos

Variable       Type         Data/Info
-------------------------------------
LabelEncoder   type         <class 'sklearn.preproces<...>ing._label.LabelEncoder'>
alcohol        DataFrame        country       beer_se<...>n\n[193 rows x 6 columns]
bigmart        DataFrame         item_identifier  ite<...>n[2444 rows x 12 columns]
df             DataFrame       person  clothes  food <...>    219      6502  354   
dt             module       <module 'datetime' from '<...>b/python3.6/datetime.py'>
iris           DataFrame         id   sepal_length_cm<...>n\n[150 rows x 6 columns]
np             module       <module 'numpy' from '/us<...>kages/numpy/__init__.py'>
num            list         n=6
olympics       DataFrame          ID     Name        <...>n[2803 rows x 13 columns]
pd             module       <module 'pandas' from '/u<...>ages/pandas/__init__.py'>
plt            module       <module 'matplotlib.pyplo<...>es/matplotlib/pyplot.py'>
pr             module       <module 'pandas_profilin

These are the variables that have been defined till now.

***
# Deleting Stored Objects

In [114]:
# del titanic

Uncomment the command and run it to remove the variable.

***
# Extracting data from Python to CSV

In [115]:
# Writing the data in current working directory
# titanic.to_csv('titanic.csv', index = False)

# Writing the data into a specified location
# titanic.to_csv('C:\\Users\\Jitendra Alim\\Desktop\\titanic', index = False)

Uncomment the command and run it to write the data.

***
# Subsetting data based on Row and Column conditions

In [116]:
titanic.loc[0:10,['passenger_id','name','gender','age']]

passenger_id name                                               gender   \
0    1                                      Braund, Mr. Owen Harris    male   
1    2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2    3                                       Heikkinen, Miss. Laina  female   
3    4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4    5                                     Allen, Mr. William Henry    male   
5    6                                             Moran, Mr. James    male   
6    7                                      McCarthy, Mr. Timothy J    male   
7    8                               Palsson, Master. Gosta Leonard    male   
8    9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
9   10                          Nasser, Mrs. Nicholas (Adele Achem)  female   
10  11                              Sandstrom, Miss. Marguerite Rut  female   

    age   
0  22.00  
1  38.00  
2  26.00  
3  35.00  
4  35.00  
5    nan  
6  54.00  
7   2.00  
8  27.00  
9  14.00  
10  4.00

***
# Obtaining initial observations of the data

In [117]:
titanic.head()

passenger_id  survived  pclass  \
0  1             0         3        
1  2             1         1        
2  3             1         3        
3  4             1         1        
4  5             0         3        

  name                                               gender   age   sib_sp  \
0                            Braund, Mr. Owen Harris    male 22.00  1        
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00  1        
2                             Heikkinen, Miss. Laina  female 26.00  0        
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00  1        
4                           Allen, Mr. William Henry    male 35.00  0        

   parch ticket             fare cabin embarked  
0  0             A/5 21171  7.25   NaN  S        
1  0              PC 17599 71.28   C85  C        
2  0      STON/O2. 3101282  7.92   NaN  S        
3  0                113803 53.10  C123  S        
4  0                373450  8.05   NaN  S

***
# Obtaining initial n observations of the data

In [118]:
alcohol.head(10)

country             beer_servings  spirit_servings  wine_servings  \
0        Afghanistan    0              0                0             
1            Albania   89            132               54             
2            Algeria   25              0               14             
3            Andorra  245            138              312             
4             Angola  217             57               45             
5  Antigua & Barbuda  102            128               45             
6          Argentina  193             25              221             
7            Armenia   21            179               11             
8          Australia  261             72              212             
9            Austria  279             75              191             

   total_litres_of_pure_alcohol continent  
0  0.00                           AS       
1  4.90                           EU       
2  0.70                           AF       
3 12.40                           EU       
4  5.90                           AF       
5  4.90                          NaN       
6  8.30                           SA       
7  3.80                           EU       
8 10.40                           OC       
9  9.70                           EU

***
# Obtaining later observations of the data

In [119]:
ufo.tail()

city       colors shape     state time            
5859  In-flight  NaN         NaN   NV    6/19/1988 5:30
5860     Woburn  NaN       LIGHT   MA    6/20/1988 1:00
5861     Conroe  NaN    FIREBALL   TX   6/20/1988 17:30
5862   Columbia  NaN        OVAL   PA   6/20/1988 20:00
5863       Napa  RED           C  NaN               NaN

***
# Obtaining later n observations of the data

In [120]:
titanic.tail(10)

passenger_id  survived  pclass name                                       \
881  882           0         3                             Markun, Mr. Johann   
882  883           0         3                   Dahlberg, Miss. Gerda Ulrika   
883  884           0         2                  Banfield, Mr. Frederick James   
884  885           0         3                         Sutehall, Mr. Henry Jr   
885  886           0         3           Rice, Mrs. William (Margaret Norton)   
886  887           0         2                          Montvila, Rev. Juozas   
887  888           1         1                   Graham, Miss. Margaret Edith   
888  889           0         3       Johnston, Miss. Catherine Helen "Carrie"   
889  890           1         1                          Behr, Mr. Karl Howell   
890  891           0         3                            Dooley, Mr. Patrick   

    gender   age   sib_sp  parch ticket             fare cabin embarked  
881    male 33.00  0       0                349257  7.90   NaN  S        
882  female 22.00  0       0                  7552 10.52   NaN  S        
883    male 28.00  0       0      C.A./SOTON 34068 10.50   NaN  S        
884    male 25.00  0       0       SOTON/OQ 392076  7.05   NaN  S        
885  female 39.00  0       5                382652 29.12   NaN  Q        
886    male 27.00  0       0                211536 13.00   NaN  S        
887  female 19.00  0       0                112053 30.00   B42  S        
888  female   nan  1       2            W./C. 6607 23.45   NaN  S        
889    male 26.00  0       0                111369 30.00  C148  C        
890    male 32.00  0       0                370376  7.75   NaN  Q

***
# Names of variables in the data

In [121]:
titanic.columns

Index(['passenger_id', 'survived', 'pclass', 'name', 'gender', 'age', 'sib_sp',
       'parch', 'ticket', 'fare', 'cabin', 'embarked'],
      dtype='object')

In [122]:
list(titanic)

['passenger_id',
 'survived',
 'pclass',
 'name',
 'gender',
 'age',
 'sib_sp',
 'parch',
 'ticket',
 'fare',
 'cabin',
 'embarked']

In [123]:
print(list(titanic))

['passenger_id', 'survived', 'pclass', 'name', 'gender', 'age', 'sib_sp', 'parch', 'ticket', 'fare', 'cabin', 'embarked']


***
# Number of rows in the data

In [124]:
len(titanic)

891

***
# Number of columns in the data

In [125]:
len(titanic.columns)

12

***
# Dimension of the data

In [126]:
titanic.shape

(891, 12)

In [127]:
print('Dimension of Titanic :', titanic.shape)

Dimension of Titanic : (891, 12)


***
# Checking the types of the variables in the data

In [128]:
titanic.dtypes

passenger_id      int64
survived          int64
pclass            int64
name             object
gender           object
age             float64
sib_sp            int64
parch             int64
ticket           object
fare            float64
cabin            object
embarked         object
dtype: object

---
# Extracting Numerical Variables from the data

In [129]:
titanic._get_numeric_data().head()

passenger_id  survived  pclass  age   sib_sp  parch  fare
0  1             0         3      22.00  1       0      7.25
1  2             1         1      38.00  1       0     71.28
2  3             1         3      26.00  0       0      7.92
3  4             1         1      35.00  1       0     53.10
4  5             0         3      35.00  0       0      8.05

In [130]:
titanic.select_dtypes(include='number').head()

passenger_id  survived  pclass  age   sib_sp  parch  fare
0  1             0         3      22.00  1       0      7.25
1  2             1         1      38.00  1       0     71.28
2  3             1         3      26.00  0       0      7.92
3  4             1         1      35.00  1       0     53.10
4  5             0         3      35.00  0       0      8.05

---
# Extracting Non-Numeric Variables from the data

In [131]:
titanic.select_dtypes(exclude='number').head()

name                                               gender   \
0                            Braund, Mr. Owen Harris    male   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                             Heikkinen, Miss. Laina  female   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                           Allen, Mr. William Henry    male   

  ticket            cabin embarked  
0         A/5 21171   NaN  S        
1          PC 17599   C85  C        
2  STON/O2. 3101282   NaN  S        
3            113803  C123  S        
4            373450   NaN  S

---
# Basic Information of the data

In [132]:
ufo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5864 entries, 0 to 5863
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   city    5853 non-null   object
 1   colors  614 non-null    object
 2   shape   5318 non-null   object
 3   state   5863 non-null   object
 4   time    5863 non-null   object
dtypes: object(5)
memory usage: 229.2+ KB


Here, we are unable to know the exact memory required for the data, since we can only see that the data exceeds 229.2 + KB, but how much exactly we don't know.

***
# Basic Information of the data with exact Memory Usage

In [133]:
ufo.info(memory_usage = 'deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5864 entries, 0 to 5863
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   city    5853 non-null   object
 1   colors  614 non-null    object
 2   shape   5318 non-null   object
 3   state   5863 non-null   object
 4   time    5863 non-null   object
dtypes: object(5)
memory usage: 1.6 MB


Now we can see the exact memory usage of the data which is 1.6 MB

***
# Checking the Memory Usage of each variable

In [134]:
ufo.memory_usage(deep = True)

Index        128
city      387911
colors    206260
shape     349924
state     345949
time      419620
dtype: int64

***
# 8-Point Summary of numeric variables

In [135]:
titanic.describe()

passenger_id  survived  pclass  age    sib_sp  parch  fare 
count 891.00        891.00    891.00  714.00 891.00  891.00 891.00
mean  446.00          0.38      2.31   29.70   0.52    0.38  32.20
std   257.35          0.49      0.84   14.53   1.10    0.81  49.69
min     1.00          0.00      1.00    0.42   0.00    0.00   0.00
25%   223.50          0.00      2.00   20.12   0.00    0.00   7.91
50%   446.00          0.00      3.00   28.00   0.00    0.00  14.45
75%   668.50          1.00      3.00   38.00   1.00    0.00  31.00
max   891.00          1.00      3.00   80.00   8.00    6.00 512.33

***
# Checking for variables with Missing Value

In [136]:
titanic.isnull().any()

passenger_id    False
survived        False
pclass          False
name            False
gender          False
age              True
sib_sp          False
parch           False
ticket          False
fare            False
cabin            True
embarked         True
dtype: bool

There are missing values in Age, Cabin and Embarked variables of Titanic dataset.

***
# Number of Missing Values in each variable

In [137]:
titanic.isnull().sum()

passenger_id      0
survived          0
pclass            0
name              0
gender            0
age             177
sib_sp            0
parch             0
ticket            0
fare              0
cabin           687
embarked          2
dtype: int64

***
# Percentage of Missing Values in each variable in a data

In [138]:
titanic.isnull().mean()*100

passenger_id    0.00
survived        0.00
pclass          0.00
name            0.00
gender          0.00
age            19.87
sib_sp          0.00
parch           0.00
ticket          0.00
fare            0.00
cabin          77.10
embarked        0.22
dtype: float64

***
# Number of rows satisfying a condition

In [139]:
(titanic['age'] > 35).sum()

217

There were 217 passengers on titanic having age more than 35 years.

***
# Percentage of rows satisfying a condition

In [140]:
(titanic['age'] > 35).mean()*100

24.354657687991022

24.35% of the passengers on titanic were more than 35 years in age.

***
# Concatenating Data Frames

In [141]:
pd.concat([titanic['name'], titanic['survived']], axis = 1).head()

name                                                survived
0                            Braund, Mr. Owen Harris  0       
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  1       
2                             Heikkinen, Miss. Laina  1       
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  1       
4                           Allen, Mr. William Henry  0

***
# Recoding Variables

In [142]:
titanic['gender_code'] = titanic['gender'].map({'male':'M', 'female':'F'})
titanic[['gender','gender_code']].head()

gender  gender_code
0    male  M         
1  female  F         
2  female  F         
3  female  F         
4    male  M

***
# Automatic Variable Encoding - Pandas Method

In [143]:
bigmart['outlet_type1'] = bigmart['outlet_type'].factorize()[0]
bigmart[['outlet_type','outlet_type1']].head(10)

outlet_type         outlet_type1
0  Supermarket Type1  0           
1  Supermarket Type2  1           
2  Supermarket Type1  0           
3      Grocery Store  2           
4  Supermarket Type1  0           
5  Supermarket Type2  1           
6  Supermarket Type1  0           
7  Supermarket Type3  3           
8  Supermarket Type1  0           
9  Supermarket Type1  0

<p> LabelEncoder encodes the variables in the alphabetical order </p>

***
# Automatic Variable Encoding - SkLearn Method

In [144]:
from sklearn.preprocessing import LabelEncoder

In [145]:
bigmart['outlet_type1'] = LabelEncoder().fit_transform(bigmart['outlet_type'].astype('str'))
bigmart[['outlet_type','outlet_type1']].head(10)

outlet_type         outlet_type1
0  Supermarket Type1  1           
1  Supermarket Type2  2           
2  Supermarket Type1  1           
3      Grocery Store  0           
4  Supermarket Type1  1           
5  Supermarket Type2  2           
6  Supermarket Type1  1           
7  Supermarket Type3  3           
8  Supermarket Type1  1           
9  Supermarket Type1  1

<p> Factorize method encodes the variable in the order which they appear </p>

***
# One Hot Encoding (Dummy Encoding)

In [146]:
pd.concat([titanic['gender'],pd.get_dummies(titanic['gender'])], axis = 1).head()

gender   female  male
0    male  0       1   
1  female  1       0   
2  female  1       0   
3  female  1       0   
4    male  0       1

***
# Changing variables to Uppercase

In [147]:
for vars in ['name','gender']:
    titanic[f'{vars}_new'] = titanic[vars].str.upper()
    
titanic[['name','name_new','gender','gender_new']]

name                                                \
0                              Braund, Mr. Owen Harris   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...   
2                               Heikkinen, Miss. Laina   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4                             Allen, Mr. William Henry   
..                                                 ...   
886                              Montvila, Rev. Juozas   
887                       Graham, Miss. Margaret Edith   
888           Johnston, Miss. Catherine Helen "Carrie"   
889                              Behr, Mr. Karl Howell   
890                                Dooley, Mr. Patrick   

    name_new                                           gender  gender_new  
0                              BRAUND, MR. OWEN HARRIS    male    MALE     
1    CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  FEMALE     
2                               HEIKKINEN, MISS. LAINA  female  FEMALE     
3         FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)  female  FEMALE     
4                             ALLEN, MR. WILLIAM HENRY    male    MALE     
..                                                 ...     ...        ...  
886                              MONTVILA, REV. JUOZAS    male    MALE     
887                       GRAHAM, MISS. MARGARET EDITH  female  FEMALE     
888           JOHNSTON, MISS. CATHERINE HELEN "CARRIE"  female  FEMALE     
889                              BEHR, MR. KARL HOWELL    male    MALE     
890                                DOOLEY, MR. PATRICK    male    MALE     

[891 rows x 4 columns]

***
# Changing variables to Lowercase

In [148]:
for vars in ['name','embarked']:
    titanic[f'{vars}_new'] = titanic[vars].str.lower()
    
titanic[['name','name_new','embarked','embarked_new']]

name                                                \
0                              Braund, Mr. Owen Harris   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...   
2                               Heikkinen, Miss. Laina   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4                             Allen, Mr. William Henry   
..                                                 ...   
886                              Montvila, Rev. Juozas   
887                       Graham, Miss. Margaret Edith   
888           Johnston, Miss. Catherine Helen "Carrie"   
889                              Behr, Mr. Karl Howell   
890                                Dooley, Mr. Patrick   

    name_new                                           embarked embarked_new  
0                              braund, mr. owen harris  S        s            
1    cumings, mrs. john bradley (florence briggs th...  C        c            
2                               heikkinen, miss. laina  S        s            
3         futrelle, mrs. jacques heath (lily may peel)  S        s            
4                             allen, mr. william henry  S        s            
..                                                 ...      ...          ...  
886                              montvila, rev. juozas  S        s            
887                       graham, miss. margaret edith  S        s            
888           johnston, miss. catherine helen "carrie"  S        s            
889                              behr, mr. karl howell  C        c            
890                                dooley, mr. patrick  Q        q            

[891 rows x 4 columns]

***
# Converting Continuous variable into Categorical bins

In [149]:
titanic['age_group'] = pd.cut(x = titanic['age'], bins = [0,18,65,99], labels = ['child','adult','elderly'], right = True)
titanic[['name','age','age_group']].head(10)

name                                                age  age_group
0                            Braund, Mr. Owen Harris 22.00  adult   
1  Cumings, Mrs. John Bradley (Florence Briggs Th... 38.00  adult   
2                             Heikkinen, Miss. Laina 26.00  adult   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel) 35.00  adult   
4                           Allen, Mr. William Henry 35.00  adult   
5                                   Moran, Mr. James   nan    NaN   
6                            McCarthy, Mr. Timothy J 54.00  adult   
7                     Palsson, Master. Gosta Leonard  2.00  child   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) 27.00  adult   
9                Nasser, Mrs. Nicholas (Adele Achem) 14.00  child

***
# Number of Unique values in a variable

In [150]:
alcohol['continent'].nunique()

5

***
# Unique values in a variable

In [151]:
alcohol['continent'].unique()

array(['AS', 'EU', 'AF', nan, 'SA', 'OC'], dtype=object)

In [152]:
print(alcohol['continent'].unique())

['AS' 'EU' 'AF' nan 'SA' 'OC']


***
# Count of unique values in a variable (w/o considering the Missing Values)

In [153]:
alcohol['continent'].isnull().sum()

23

There are missing values in Continent variable of Alcohol dataset

In [154]:
alcohol['continent'].value_counts()

AF    53
EU    45
AS    44
OC    16
SA    12
Name: continent, dtype: int64

This counts does not represent the entire dataset as we have not accounted for the missing values in Continent variable of Alcohol dataset.

In [155]:
alcohol['continent'].value_counts(dropna = False)

AF     53
EU     45
AS     44
NaN    23
OC     16
SA     12
Name: continent, dtype: int64

***
# Percentage of unique values in a variable (w/o considering Missing Values)

In [156]:
alcohol['continent'].value_counts(normalize = True)

AF   0.31
EU   0.26
AS   0.26
OC   0.09
SA   0.07
Name: continent, dtype: float64

***
# Percentage of unique values in a variable (considering Missing Values)

In [157]:
alcohol['continent'].value_counts(dropna = False, normalize = True)

AF    0.27
EU    0.23
AS    0.23
NaN   0.12
OC    0.08
SA    0.06
Name: continent, dtype: float64

We get False percentages when we don't consider the missing values.

***
# Special HTML Summary Report of the data

In [158]:
# pr.ProfileReport(titanic)

<p> Profile Report is one of the awesome exploratory report </p>

***
# Importing Tips

1. Only read in columns you need.
2. Use 'category' data type with categorical data.
3. Use 'int' data type for strictly integer data.

***
# Importing data specifying variable type for single variable

In [159]:
alcohol = pd.read_csv('http://bit.ly/alcohol_consumed', dtype = {'beer_servings':'float'})
alcohol.dtypes

country                          object
beer_servings                   float64
spirit_servings                   int64
wine_servings                     int64
total_litres_of_pure_alcohol    float64
continent                        object
dtype: object

***
# Importing data specifying multiple variable types for multiple variables

In [160]:
alcohol = pd.read_csv('http://bit.ly/alcohol_consumed', dtype = {'beer_servings':'float','country':'category'})
alcohol.dtypes

country                         category
beer_servings                    float64
spirit_servings                    int64
wine_servings                      int64
total_litres_of_pure_alcohol     float64
continent                         object
dtype: object

***
# Importing data specifying Date variables

In [161]:
ufo = pd.read_csv('http://bit.ly/ufo_sights', parse_dates = ['time'])
ufo.dtypes

city              object
colors            object
shape             object
state             object
time      datetime64[ns]
dtype: object

Since, we used parse_dates argument while importing the data, the datatype of Time vairable is datetime variable or else by default it would have just been a variable of datatype as object.

This will reduce the memory usage by Time variable.

There are more special benifits of using Date and Time related vairable as datetime variable, which will be covered further.

***
# Renaming a single variable

In [162]:
titanic.columns

Index(['passenger_id', 'survived', 'pclass', 'name', 'gender', 'age', 'sib_sp',
       'parch', 'ticket', 'fare', 'cabin', 'embarked', 'gender_code',
       'name_new', 'gender_new', 'embarked_new', 'age_group'],
      dtype='object')

In [163]:
titanic = titanic.rename(columns = {'pclass':'class'})
titanic.columns

Index(['passenger_id', 'survived', 'class', 'name', 'gender', 'age', 'sib_sp',
       'parch', 'ticket', 'fare', 'cabin', 'embarked', 'gender_code',
       'name_new', 'gender_new', 'embarked_new', 'age_group'],
      dtype='object')

***
# Renaming multiple variables

In [164]:
bigmart.columns

Index(['item_identifier', 'item_weight', 'item_fat_content', 'item_visibility',
       'item_type', 'item_mrp', 'outlet_identifier',
       'outlet_establishment_year', 'outlet_size', 'outlet_location_type',
       'outlet_type', 'item_outlet_sales', 'outlet_type1'],
      dtype='object')

In [165]:
bigmart = bigmart.rename(columns = {'item_identifier':'item','item_type':'type'})
bigmart.head()

item    item_weight item_fat_content  item_visibility  \
0  FDA15  9.30         Low Fat         0.02               
1  DRC01  5.92         Regular         0.02               
2  FDN15 17.50         Low Fat         0.02               
3  FDX07 19.20         Regular         0.00               
4  NCD19  8.93         Low Fat         0.00               

  type                    item_mrp outlet_identifier  \
0                  Dairy 249.81     OUT049             
1            Soft Drinks  48.27     OUT018             
2                   Meat 141.62     OUT049             
3  Fruits and Vegetables 182.09     OUT010             
4              Household  53.86     OUT013             

   outlet_establishment_year outlet_size outlet_location_type  \
0 1,999.00                    Medium      Tier 1                
1 2,009.00                    Medium      Tier 3                
2 1,999.00                    Medium      Tier 1                
3 1,998.00                       NaN      Tier 3                
4 1,987.00                      High      Tier 3                

  outlet_type         item_outlet_sales  outlet_type1  
0  Supermarket Type1 3,735.14            1             
1  Supermarket Type2   443.42            2             
2  Supermarket Type1 2,097.27            1             
3      Grocery Store   732.38            0             
4  Supermarket Type1   994.71            1

***
# Converting a variable into categorical variable

In [166]:
print('Datatype of Item_Fat_Content     :', bigmart['item_fat_content'].dtypes)
print('Memory Usage of Item_Fat_Content :', bigmart['item_fat_content'].memory_usage(deep = True))

Datatype of Item_Fat_Content     : object
Memory Usage of Item_Fat_Content : 155972


In [167]:
bigmart['item_fat_content'] = bigmart['item_fat_content'].astype('category')
print('Datatype of Item_Fat_Content     :', bigmart['item_fat_content'].dtypes)
print('Memory Usage of Item_Fat_Content :', bigmart['item_fat_content'].memory_usage(deep = True))

Datatype of Item_Fat_Content     : category
Memory Usage of Item_Fat_Content : 3043


Converting a variable into categorical variable reduced the memory usage of that variable and hence reduces the size of the data frame. This reduction of the size of the data frame allows more faster calculation and execution of codes.

***
# Selecting only numeric columns from data

In [168]:
# Method 1 (Long Method)
num = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
bigmart.select_dtypes(include = num).head()

item_weight  item_visibility  item_mrp  outlet_establishment_year  \
0  9.30        0.02             249.81    1,999.00                     
1  5.92        0.02              48.27    2,009.00                     
2 17.50        0.02             141.62    1,999.00                     
3 19.20        0.00             182.09    1,998.00                     
4  8.93        0.00              53.86    1,987.00                     

   item_outlet_sales  outlet_type1  
0 3,735.14            1             
1   443.42            2             
2 2,097.27            1             
3   732.38            0             
4   994.71            1

In [169]:
# Method 2
bigmart.select_dtypes(include = 'number').head()

item_weight  item_visibility  item_mrp  outlet_establishment_year  \
0  9.30        0.02             249.81    1,999.00                     
1  5.92        0.02              48.27    2,009.00                     
2 17.50        0.02             141.62    1,999.00                     
3 19.20        0.00             182.09    1,998.00                     
4  8.93        0.00              53.86    1,987.00                     

   item_outlet_sales  outlet_type1  
0 3,735.14            1             
1   443.42            2             
2 2,097.27            1             
3   732.38            0             
4   994.71            1

In [170]:
# Method 3
bigmart._get_numeric_data().head()

item_weight  item_visibility  item_mrp  outlet_establishment_year  \
0  9.30        0.02             249.81    1,999.00                     
1  5.92        0.02              48.27    2,009.00                     
2 17.50        0.02             141.62    1,999.00                     
3 19.20        0.00             182.09    1,998.00                     
4  8.93        0.00              53.86    1,987.00                     

   item_outlet_sales  outlet_type1  
0 3,735.14            1             
1   443.42            2             
2 2,097.27            1             
3   732.38            0             
4   994.71            1

***
# Drop variables with any Missing Values

In [171]:
print('Dimension of Titanic :', titanic.shape)

Dimension of Titanic : (891, 17)


In [172]:
titanic.dropna(axis='columns')

passenger_id  survived  class  \
0      1           0         3       
1      2           1         1       
2      3           1         3       
3      4           1         1       
4      5           0         3       
..            ...       ...    ...   
886  887           0         2       
887  888           1         1       
888  889           0         3       
889  890           1         1       
890  891           0         3       

    name                                               gender   sib_sp  parch  \
0                              Braund, Mr. Owen Harris    male  1       0       
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  1       0       
2                               Heikkinen, Miss. Laina  female  0       0       
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  1       0       
4                             Allen, Mr. William Henry    male  0       0       
..                                                 ...     ...     ...    ...   
886                              Montvila, Rev. Juozas    male  0       0       
887                       Graham, Miss. Margaret Edith  female  0       0       
888           Johnston, Miss. Catherine Helen "Carrie"  female  1       2       
889                              Behr, Mr. Karl Howell    male  0       0       
890                                Dooley, Mr. Patrick    male  0       0       

    ticket             fare gender_code  \
0           A/5 21171  7.25  M            
1            PC 17599 71.28  F            
2    STON/O2. 3101282  7.92  F            
3              113803 53.10  F            
4              373450  8.05  M            
..                ...   ...         ...   
886            211536 13.00  M            
887            112053 30.00  F            
888        W./C. 6607 23.45  F            
889            111369 30.00  M            
890            370376  7.75  M            

    name_new                                           gender_new  
0                              braund, mr. owen harris    MALE     
1    cumings, mrs. john bradley (florence briggs th...  FEMALE     
2                               heikkinen, miss. laina  FEMALE     
3         futrelle, mrs. jacques heath (lily may peel)  FEMALE     
4                             allen, mr. william henry    MALE     
..                                                 ...        ...  
886                              montvila, rev. juozas    MALE     
887                       graham, miss. margaret edith  FEMALE     
888           johnston, miss. catherine helen "carrie"  FEMALE     
889                              behr, mr. karl howell    MALE     
890                                dooley, mr. patrick    MALE     

[891 rows x 12 columns]

Since, 3 variables having missing values are dropped, the resultant data has 9 variables instead of original 12 variables.

***
# Dropping variables with more than 10% of Missing Values

In [173]:
titanic.dropna(thresh = len(titanic)*0.9, axis = 'columns')

passenger_id  survived  class  \
0      1           0         3       
1      2           1         1       
2      3           1         3       
3      4           1         1       
4      5           0         3       
..            ...       ...    ...   
886  887           0         2       
887  888           1         1       
888  889           0         3       
889  890           1         1       
890  891           0         3       

    name                                               gender   sib_sp  parch  \
0                              Braund, Mr. Owen Harris    male  1       0       
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  1       0       
2                               Heikkinen, Miss. Laina  female  0       0       
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  1       0       
4                             Allen, Mr. William Henry    male  0       0       
..                                                 ...     ...     ...    ...   
886                              Montvila, Rev. Juozas    male  0       0       
887                       Graham, Miss. Margaret Edith  female  0       0       
888           Johnston, Miss. Catherine Helen "Carrie"  female  1       2       
889                              Behr, Mr. Karl Howell    male  0       0       
890                                Dooley, Mr. Patrick    male  0       0       

    ticket             fare embarked gender_code  \
0           A/5 21171  7.25  S        M            
1            PC 17599 71.28  C        F            
2    STON/O2. 3101282  7.92  S        F            
3              113803 53.10  S        F            
4              373450  8.05  S        M            
..                ...   ...      ...         ...   
886            211536 13.00  S        M            
887            112053 30.00  S        F            
888        W./C. 6607 23.45  S        F            
889            111369 30.00  C        M            
890            370376  7.75  Q        M            

    name_new                                           gender_new embarked_new  
0                              braund, mr. owen harris    MALE     s            
1    cumings, mrs. john bradley (florence briggs th...  FEMALE     c            
2                               heikkinen, miss. laina  FEMALE     s            
3         futrelle, mrs. jacques heath (lily may peel)  FEMALE     s            
4                             allen, mr. william henry    MALE     s            
..                                                 ...        ...          ...  
886                              montvila, rev. juozas    MALE     s            
887                       graham, miss. margaret edith  FEMALE     s            
888           johnston, miss. catherine helen "carrie"  FEMALE     s            
889                              behr, mr. karl howell    MALE     c            
890                                dooley, mr. patrick    MALE     q            

[891 rows x 14 columns]

Since 1 variable had less than 10% of missing values, only 2 variables having more than 10% of missing values were dropped.

***
# Dropping observations with any Missing Value

In [174]:
print('Dimension of Titanic :', titanic.shape)

Dimension of Titanic : (891, 17)


In [175]:
titanic.dropna(axis = 'rows')

passenger_id  survived  class  \
1      2           1         1       
3      4           1         1       
6      7           0         1       
10    11           1         3       
11    12           1         1       
..            ...       ...    ...   
871  872           1         1       
872  873           0         1       
879  880           1         1       
887  888           1         1       
889  890           1         1       

    name                                               gender   age   sib_sp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00  1        
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00  1        
6                              McCarthy, Mr. Timothy J    male 54.00  0        
10                     Sandstrom, Miss. Marguerite Rut  female  4.00  1        
11                            Bonnell, Miss. Elizabeth  female 58.00  0        
..                                                 ...     ...   ...     ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female 47.00  1        
872                           Carlsson, Mr. Frans Olof    male 33.00  0        
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female 56.00  0        
887                       Graham, Miss. Margaret Edith  female 19.00  0        
889                              Behr, Mr. Karl Howell    male 26.00  0        

     parch ticket     fare cabin        embarked gender_code  \
1    0      PC 17599 71.28          C85  C        F            
3    0        113803 53.10         C123  S        F            
6    0         17463 51.86          E46  S        M            
10   1       PP 9549 16.70           G6  S        F            
11   0        113783 26.55         C103  S        F            
..     ...       ...   ...          ...      ...         ...   
871  1         11751 52.55          D35  S        F            
872  0           695  5.00  B51 B53 B55  S        M            
879  1         11767 83.16          C50  C        F            
887  0        112053 30.00          B42  S        F            
889  0        111369 30.00         C148  C        M            

    name_new                                           gender_new  \
1    cumings, mrs. john bradley (florence briggs th...  FEMALE      
3         futrelle, mrs. jacques heath (lily may peel)  FEMALE      
6                              mccarthy, mr. timothy j    MALE      
10                     sandstrom, miss. marguerite rut  FEMALE      
11                            bonnell, miss. elizabeth  FEMALE      
..                                                 ...        ...   
871   beckwith, mrs. richard leonard (sallie monypeny)  FEMALE      
872                           carlsson, mr. frans olof    MALE      
879      potter, mrs. thomas jr (lily alexenia wilson)  FEMALE      
887                       graham, miss. margaret edith  FEMALE      
889                              behr, mr. karl howell    MALE      

    embarked_new age_group  
1    c            adult     
3    s            adult     
6    s            adult     
10   s            child     
11   s            adult     
..           ...       ...  
871  s            adult     
872  s            adult     
879  c            adult     
887  s            adult     
889  c            adult     

[183 rows x 17 columns]

Since only these 183 observations had no missing values for any variable, they are present in the resultant data.

***
# Wide to Long Data

In [176]:
df = pd.DataFrame({'person':[1,2,3], 'clothes':[346,687,219], 'food':[3540,8740,6502], 'travel':[198,216,354]})
df.head()

person  clothes  food  travel
0  1       346      3540  198   
1  2       687      8740  216   
2  3       219      6502  354

In [177]:
df.melt(id_vars = 'person', var_name = 'expenditure', value_name = 'expense')

person expenditure  expense
0  1       clothes      346   
1  2       clothes      687   
2  3       clothes      219   
3  1          food     3540   
4  2          food     8740   
5  3          food     6502   
6  1        travel      198   
7  2        travel      216   
8  3        travel      354

***
# Counting words in a variable

In [178]:
titanic['name_words'] = titanic['name'].str.count(' ') + 1
titanic[['name','name_words']].head()

name                                                name_words
0                            Braund, Mr. Owen Harris  4         
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  7         
2                             Heikkinen, Miss. Laina  3         
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  7         
4                           Allen, Mr. William Henry  4

***
# Pivot Tables (Single Variable, Single Aggregate Function)

In [179]:
alcohol.pivot_table(columns = 'continent')

continent                     AF    AS    EU     OC    SA   
beer_servings                61.47 37.05 193.78 89.69 175.08
spirit_servings              16.34 60.84 132.56 58.44 114.75
total_litres_of_pure_alcohol  3.01  2.17   8.62  3.38   6.31
wine_servings                16.26  9.07 142.22 35.62  62.42

The default aggregate function of Pivot Tables is **Mean**

In [180]:
alcohol.pivot_table(columns = 'continent', aggfunc = 'mean')

continent                     AF    AS    EU     OC    SA   
beer_servings                61.47 37.05 193.78 89.69 175.08
spirit_servings              16.34 60.84 132.56 58.44 114.75
total_litres_of_pure_alcohol  3.01  2.17   8.62  3.38   6.31
wine_servings                16.26  9.07 142.22 35.62  62.42

In [181]:
alcohol.pivot_table(index = 'continent', aggfunc = 'sum')

beer_servings  spirit_servings  total_litres_of_pure_alcohol  \
continent                                                                 
AF        3,258.00         866            159.40                          
AS        1,630.00        2677             95.50                          
EU        8,720.00        5965            387.80                          
OC        1,435.00         935             54.10                          
SA        2,101.00        1377             75.70                          

           wine_servings  
continent                 
AF          862           
AS          399           
EU         6400           
OC          570           
SA          749

In [182]:
alcohol.pivot_table(index = 'continent', aggfunc = 'sum').reset_index()

continent  beer_servings  spirit_servings  total_litres_of_pure_alcohol  \
0  AF       3,258.00         866            159.40                          
1  AS       1,630.00        2677             95.50                          
2  EU       8,720.00        5965            387.80                          
3  OC       1,435.00         935             54.10                          
4  SA       2,101.00        1377             75.70                          

   wine_servings  
0   862           
1   399           
2  6400           
3   570           
4   749

***
# Pivot Tables (Single Variable, Single Aggregate Function) with Margins

In [183]:
alcohol.pivot_table(index = 'continent', aggfunc = 'sum', margins = True, margins_name = 'Total')

beer_servings  spirit_servings  total_litres_of_pure_alcohol  \
continent                                                                 
AF         3,258.00         866           159.40                          
AS         1,630.00        2677            95.50                          
EU         8,720.00        5965           387.80                          
OC         1,435.00         935            54.10                          
SA         2,101.00        1377            75.70                          
Total     17,144.00       11820           772.50                          

           wine_servings  
continent                 
AF          862           
AS          399           
EU         6400           
OC          570           
SA          749           
Total      8980

***
# Pivot Tables (Single Variable, Multiple Aggregate Functions)

In [184]:
alcohol.pivot_table(columns = 'continent', aggfunc = ['mean','sum'])

mean                            sum       \
continent                    AF    AS    EU     OC    SA     AF         
beer_servings                61.47 37.05 193.78 89.69 175.08 3,258.00   
spirit_servings              16.34 60.84 132.56 58.44 114.75   866.00   
total_litres_of_pure_alcohol  3.01  2.17   8.62  3.38   6.31   159.40   
wine_servings                16.26  9.07 142.22 35.62  62.42   862.00   

                                                                  
continent                    AS       EU       OC       SA        
beer_servings                1,630.00 8,720.00 1,435.00 2,101.00  
spirit_servings              2,677.00 5,965.00   935.00 1,377.00  
total_litres_of_pure_alcohol    95.50   387.80    54.10    75.70  
wine_servings                  399.00 6,400.00   570.00   749.00

***
# Group By (Single Variable, Single Aggregate Function)

In [185]:
alcohol.groupby(by = 'continent').agg({'beer_servings':'mean'})

beer_servings
continent               
AF         61.47        
AS         37.05        
EU        193.78        
OC         89.69        
SA        175.08

This above groupby command creats Multi-level Index. Also, the variable names created are uninformative.

***
# Group By (Single Variable, Single Aggregate Function) with Informative Names

In [186]:
alcohol.groupby(by = 'continent').beer_servings.agg(avg_beer_serving = 'mean')

avg_beer_serving
continent                  
AF         61.47           
AS         37.05           
EU        193.78           
OC         89.69           
SA        175.08

To avoid the Muli-level Index, one can use the Index Resetting command.

In [187]:
alcohol.groupby(by = 'continent').beer_servings.agg(avg_beer_serving = 'mean').reset_index()

continent  avg_beer_serving
0  AF        61.47           
1  AS        37.05           
2  EU       193.78           
3  OC        89.69           
4  SA       175.08

***
# Group By (Single Variables, Multiple Aggregate Function)

In [188]:
alcohol.groupby(by = 'continent').agg({'beer_servings':['sum','mean']})

beer_servings       
          sum           mean  
continent                     
AF        3,258.00       61.47
AS        1,630.00       37.05
EU        8,720.00      193.78
OC        1,435.00       89.69
SA        2,101.00      175.08

In [189]:
alcohol.groupby(by = 'continent').agg({'beer_servings':['sum','mean']}).reset_index()

continent beer_servings       
            sum           mean  
0  AF       3,258.00       61.47
1  AS       1,630.00       37.05
2  EU       8,720.00      193.78
3  OC       1,435.00       89.69
4  SA       2,101.00      175.08

For Multiple Aggregate Function Group By command, Index Resetting command cannot handle Multi-level Indexing.

***
# Group By (Single Variable, Multiple Aggregate Function) with Informative Names

In [190]:
alcohol.groupby(by = 'continent').beer_servings.agg(total_servings = 'sum', avg_servings = 'mean')

total_servings  avg_servings
continent                              
AF        3,258.00         61.47       
AS        1,630.00         37.05       
EU        8,720.00        193.78       
OC        1,435.00         89.69       
SA        2,101.00        175.08

In [191]:
alcohol.groupby(by = 'continent').beer_servings.agg(total_servings = 'sum', avg_servings = 'mean').reset_index()

continent  total_servings  avg_servings
0  AF       3,258.00         61.47       
1  AS       1,630.00         37.05       
2  EU       8,720.00        193.78       
3  OC       1,435.00         89.69       
4  SA       2,101.00        175.08

Since, the names of the variables after aggregation was specified, Index Resetting command could handle Multi-level Index.

***
# Making Date from different variables

In [192]:
time = pd.DataFrame([[25,12,2019],[28,11,2019]], columns = ['day','month','year'])
time

day  month  year
0  25   12     2019
1  28   11     2019

In [193]:
time['date'] = pd.to_datetime(time[['day','month','year']])
time

day  month  year date      
0  25   12     2019 2019-12-25
1  28   11     2019 2019-11-28

***
# Sharing Code with Link

In [194]:
# %pastebin 1 - 100

<p> The above code will generate a sharable web link to share the codes from line number 1 to 100 </p>

***
# Ignoring Warnings

In [195]:
# import warning
# warning.filterwanings('ignore')

---
# Importing Large Data with Row Skipping

In [196]:
olympics = pd.read_csv('http://bit.ly/olympics_120years')
print('Dimension of Olympics :', olympics.shape)
print('Memory Usage : %.2f MB' % (olympics.memory_usage(deep = True).sum()/(1024*1024)))

Dimension of Olympics : (271116, 13)
Memory Usage : 141.30 MB


In [197]:
# Execution of this code will take time due to the use of lambda function 
olympics = pd.read_csv('http://bit.ly/olympics_120years',
                       skiprows = lambda x: x > 0 and np.random.rand() > 0.01)

print('Dimension of Olympics :', olympics.shape)
print('Memory Usage : %.2f MB' % (olympics.memory_usage(deep = True).sum()/(1024*1024)))

Dimension of Olympics : (2733, 13)
Memory Usage : 1.42 MB


## How it works

- skiprows accepts a function that is evaluated against the integer index
- x > 0 ensures that the header row is not skipped
- np.random.rand() > 0.01 returns True 99% of the times, thus skipping 99% of the data

This method can be used to write code on a representative of a large data when struggling with data loading on slower processors